### PR TITLE
fix(codex): repair legacy managed account bindings

### LINKED ISSUE #6969: [Bug][v3.20.1] Codex 官方卡 authBinding.accountId 未随账号主键变更迁移（仍为旧 workspace ID），报「账号不存在」并死锁，无法切换任何供应商
### 环境

- 版本：v3.20.0 → v3.20.1（升级后出现问题）
- 系统：Windows 11 x64
- 场景：Codex 官方卡（`providerType: codex_oauth`，绑定托管 ChatGPT 账号），绑定关系创建于 v3.20.0

### 现象

升级到 v3.20.1 后，这张绑定了托管 ChatGPT 账号的 Codex 官方卡出现死锁：

1. 卡片显示「绑定的账号不可用」和「会话已过期」；点卡片上的刷新/重新登录按钮报错：
   ```
   账号不存在: <uuid>
   ```
2. 编辑该供应商并保存，报「更新供应商失败： 账号不存在： <uuid>」。
3. 因为该卡是**当前启用中**的供应商，尝试切换到其他任何供应商（包括第三方卡）也报「切换失败： 账号不存在： <uuid>」——整个 Codex 应用的切换被完全卡死。

关键诊断特征：报错里的 `<uuid>` 不是账号库中任何账号的本地 `account_id`，而恰好等于该账号的 `chatgpt_account_id`（workspace ID）字段值。

### 根因分析（读源码确认）

- v3.20.0 的 Codex OAuth 多账号功能以 **workspace ID（`chatgpt_account_id`）作为托管账号主键**，此时 provider `meta.authBinding.accountId` 存 workspace ID 是自洽的。
- v3.20.1 修复 #2245 后，`codex_oauth_auth.json` 改为**本地 UUID（`account_id`）作主键**，`chatgpt_account_id` 仅作为上游 workspace 标识保留。
- 但 **v17→v18 数据库迁移（`migrate_v17_to_v18`）只给 `session_log_sync` 加了两列，没有重写 `providers.meta` 里的 `authBinding.accountId`**。升级说明中的「已有托管账号需要一次性重新登录」只覆盖了账号库本身，供应商侧绑定被遗漏。
- 运行时也无兜底：`Provider::managed_account_id_for()` 原样返回存储的 `accountId`，`apply_codex_official_auth()` 与切走流程的 `prepare_outgoing_managed_codex_live_auth()` 都按该 ID 查 `CodexOAuthManager`，查不到即 `CodexOAuthError::AccountNotFound`。
- 死锁扩大化的原因：`switch_normal()` 中只要 current 或 target 任一为托管绑定的 Codex 卡就启用 managed transaction，outgoing 账号解析失败 → 连「切走」也被拒绝。

即：**v3.20.0 创建的托管绑定，升级 v3.20.1 后必然全部失效**，且当前卡为绑定卡时应用整体无法切换。

### 临时解决办法

> ⚠️ 操作前先完全退出 CC Switch（含托盘进程），并备份 `~/.cc-switch/cc-switch.db`（`backups/` 目录里也有自动备份可兜底）。

1. 打开 `~/.cc-switch/codex_oauth_auth.json`，找到现存账号的本地 `account_id`（形如 UUID；注意不是 `chatgpt_account_id` 字段）。
2. 执行（需带 JSON1 的 sqlite3）：

```sql
UPDATE providers
SET meta = json_set(
    meta,
    '$.authBinding.accountId',
    '<codex_oauth_auth.json 中该账号的本地 account_id>'
)
WHERE app_type = 'codex'
  AND json_extract(meta, '$.authBinding.accountId') = '<旧 workspace ID，即报错里出现的那个 UUID>';
```

3. 重启应用。若卡片仍提示「会话已过期」，在账号管理里对该账号重新登录一次即可。

### 建议的修复方向

1. **补迁移**：新增迁移（或发一个 v3.20.x 补丁版）扫描 `providers.meta`，对 `authBinding.authProvider = "codex_oauth"` 的行，按「`chatgpt_account_id` → 现本地 `account_id`」的映射重写 `authBinding.accountId`；找不到映射时清空绑定并降级为未绑定官方卡，避免死锁。
2. **运行时兜底**：`managed_account_id_for()` 解析不到时，尝试按 `chatgpt_account_id` 匹配一次并自动改写绑定（幂等自愈）。
3. **交互兜底**：切换/更新遇到 `AccountNotFound` 时，给出「绑定账号已失效，请重新选择账号」的可操作引导，而不是让 current 卡阻塞整个应用的切换（切走路径可考虑对 outgoing 账号缺失做降级清理而非直接失败）。

### PR DESCRIPTION
**Pull request by allenxu09**
## Summary
- detect legacy Codex OAuth provider bindings that still store a ChatGPT workspace ID
- rewrite only bindings that uniquely map to a ready local managed account
- leave shared or otherwise ambiguous workspaces untouched to avoid cross-account binding

## Verification
- `cargo test --manifest-path src-tauri/Cargo.toml legacy_binding_aliases_only_include_unique_ready_accounts`
- `cargo test --manifest-path src-tauri/Cargo.toml repairs_legacy_codex_oauth_binding_once`
- `cargo fmt --manifest-path src-tauri/Cargo.toml --all -- --check`
- `cargo clippy --manifest-path src-tauri/Cargo.toml --all-targets -- -D warnings`

Closes #6969

### CODE REVIEW

**Comment by farion1231** [comment]
@codex review

**Review by farion1231 — COMMENTED** [review]
Thanks — the overall shape here is right: a startup-time idempotent repair (instead of a schema migration) retries on every launch, doesn't bump the schema version, and a repair failure only warns instead of blocking startup. Merge against latest main is clean and the full suite passes. Two issues before this can merge, both small and both inside the two functions you added:

**[P1] Unready accounts must still veto workspace uniqueness — otherwise a shared workspace can be silently rebound to a different user.**

In [`legacy_binding_account_id_aliases`](https://github.com/farion1231/cc-switch/blob/c9062b68c6e30b9f809c0e85df6cef21f3864349/src-tauri/src/proxy/providers/codex_oauth_auth.rs#L1358-L1374), an account without a usable `id_token` identity is `continue`d past, so it never participates in the ambiguity vote. Concrete failure: user A's legacy account (keyed by the workspace ID, quarantined after the upgrade) is still in the store; user B then signs in to the same shared workspace and gets a ready UUID account. The workspace now looks uniquely owned by B, and startup rewrites A's provider binding to B's local account — A's card silently starts using B's credentials. (The inline bot comment on this hunk flags the same scenario.)

Two things make the conservative fix strictly better:

- A binding whose stored ID still resolves (the workspace-keyed legacy account is still present) is **not** deadlocked — the card shows "re-login" and targeted reauth restores it in place, preserving the binding. The repair shouldn't touch resolvable bindings at all.
- Counting unready accounts as ambiguity loses **zero** deadlock coverage: the deadlock requires the bound ID to be absent from the store, and an absent account casts no vote either way.

Minimal fix — turn the skip into a veto:

```rust
if account
    .id_token
    .as_deref()
    .and_then(crate::codex_config::extract_codex_id_token_user_identity)
    .is_none()
{
    // Unready accounts can't be rewrite targets, but they must still veto
    // uniqueness: the workspace may belong to their (unproven) owner.
    candidates
        .entry(workspace_id.to_string())
        .and_modify(|candidate| *candidate = None)
        .or_insert(None);
    continue;
}
```

`legacy_binding_aliases_only_include_unique_ready_accounts` currently locks in the unsafe behavior (the stale workspace-keyed account is ignored and the alias is still produced) — its assertion needs to flip: with the stale account present, that workspace must yield no alias.

**[P2] The repair only scans `app_type = 'codex'`, but Claude and Claude Desktop providers carry the same legacy bindings.**

Both the [query](https://github.com/farion1231/cc-switch/blob/c9062b68c6e30b9f809c0e85df6cef21f3864349/src-tauri/src/database/dao/providers.rs#L196) and the [update](https://github.com/farion1231/cc-switch/blob/c9062b68c6e30b9f809c0e85df6cef21f3864349/src-tauri/src/database/dao/providers.rs#L238) are pinned to `'codex'`. But `claude` and `claude-desktop` providers have shipped `providerType: "codex_oauth"` reverse-proxy presets since v3.20.0, and their `authBinding.accountId` stores the same legacy workspace ID. After the upgrade those bindings dangle too: in the [forwarder](https://github.com/farion1231/cc-switch/blob/21fda0ea0/src-tauri/src/proxy/forwarder.rs#L1750-L1763), a dangling binding fails `get_valid_token_for_account` on every request — the `default_account_id` fallback only applies when the binding is `None`, not when it's stale. Fix: drop the app filter from the SELECT (the `authProvider == "codex_oauth"` predicate already keeps copilot/xai bindings out), read `app_type` per row, update with the composite key `WHERE id = ?2 AND app_type = ?3`, and add a claude-row test.

**Follow-ups (fine to defer to separate PRs):**

- When no alias applies (ambiguous workspace, or the account was deleted), switching away from the bound Codex card still hard-fails on `AccountNotFound` — degrading that path (skip adoption or clear the dangling binding) would close the remaining deadlock, per suggestion 3 in #6969.
- The repair only runs at startup: a user who re-authenticates after launch needs one more restart. Triggering the repair once after a successful login would remove that.
- One malformed `meta` JSON row aborts the whole repair; per-row tolerance (skip and warn) matches how other meta rewrites in the codebase handle it.

Fix the P1 and P2 and this is good to merge — roughly 15 lines total plus test adjustments. Happy to re-review quickly.

File: src-tauri/src/proxy/providers/codex_oauth_auth.rs [hunk 1]
```diff
@@ -1348,6 +1348,42 @@ impl CodexOAuthManager {
...
+            let Some(workspace_id) = account.chatgpt_account_id.as_deref() else {
+                continue;
+            };
+            if account
+                .id_token
+                .as_deref()
+                .and_then(crate::codex_config::extract_codex_id_token_user_identity)
+                .is_none()
+            {
+                continue;
```
[the remark this answers is not in the document]
Comment (allenxu09) [reply src-tauri/src/proxy/providers/codex_oauth_auth.rs:1372]: Fixed in 15cb3e3e: accounts without a usable id_token now mark their workspace as ambiguous, while an uncontested ready account still produces an alias. The regression test covers both cases.

**Comment by allenxu09** [comment]
Addressed both required review items in 15cb3e3e: unready accounts now veto workspace alias uniqueness, and the repair scans all provider app types while updating by the full (id, app_type) key. The regression test now covers both codex and claude rows plus contested and uncontested workspaces. Verified with cargo test --manifest-path src-tauri/Cargo.toml binding (8 passed), cargo fmt --manifest-path src-tauri/Cargo.toml --all -- --check, and OCR delegate review of all 3 PR files (100% coverage, no remaining findings). Ready for re-review.

**Review by farion1231 — COMMENTED** [review]
Thanks for the quick turnaround. The cross-app half is verified clean on my side: the all-provider scan with the `source`/`authProvider` predicates, the composite-key update, and the codex+claude regression test closing the loop through a real `save_provider`. Merged against latest main, the full suite passes here (2892 Rust tests, frontend typecheck + 1071/1071), and CI is green on all four backend platforms.

One item remains — and it's partly on me: the snippet in my last review only covered one of the two skip branches my own writeup described, and 15cb3e3e implements that snippet faithfully.

**[P1] Real quarantined v1 records still cast no veto — they have no `chatgpt_account_id` at all.**

The new veto sits on the second branch (workspace present, identity unprovable). But an account quarantined by the v3.20.1 upgrade takes the first branch at [codex_oauth_auth.rs#L1359-L1361](https://github.com/farion1231/cc-switch/blob/15cb3e3efe7f5fd2787b00fc2eb8d11ea8ca2edf/src-tauri/src/proxy/providers/codex_oauth_auth.rs#L1359-L1361) and is skipped without voting: the v1 store format had no `chatgpt_account_id` field (the map key *was* the workspace ID), so serde loads it as `None`. That's exactly the shape [`test_v1_store_stays_quarantined_until_targeted_reauth`](https://github.com/farion1231/cc-switch/blob/15cb3e3efe7f5fd2787b00fc2eb8d11ea8ca2edf/src-tauri/src/proxy/providers/codex_oauth_auth.rs#L2487) locks in, and what the comment at [L294](https://github.com/farion1231/cc-switch/blob/15cb3e3efe7f5fd2787b00fc2eb8d11ea8ca2edf/src-tauri/src/proxy/providers/codex_oauth_auth.rs#L294) documents: "A missing workspace marks a quarantined pre-v2 record."

Reproduced in a merge-tree worktree: insert a quarantined record keyed `legacy-workspace` (`chatgpt_account_id: None`) plus a ready account `local-b` on that same workspace — `legacy_binding_account_id_aliases` still returns `{"legacy-workspace": "local-b"}`, and startup rewrites user A's still-resolvable binding to user B's account. The stale account in [`legacy_binding_aliases_only_include_uncontested_ready_accounts`](https://github.com/farion1231/cc-switch/blob/15cb3e3efe7f5fd2787b00fc2eb8d11ea8ca2edf/src-tauri/src/proxy/providers/codex_oauth_auth.rs#L2259) is built with the test helper, which always sets `chatgpt_account_id = Some(...)` — so the test is green against a shape that doesn't occur in upgraded stores.

Minimal fix, verified locally (the quarantined-shape repro turns green; all 8 existing binding tests stay green): veto any workspace ID that is still a live local account key — a binding storing such an ID resolves, so it isn't deadlocked and must not be rewritten. In the final filter at [L1385](https://github.com/farion1231/cc-switch/blob/15cb3e3efe7f5fd2787b00fc2eb8d11ea8ca2edf/src-tauri/src/proxy/providers/codex_oauth_auth.rs#L1385):

```rust
candidates
    .into_iter()
    .filter_map(|(workspace_id, local_id)| {
        local_id
            .filter(|local_id| local_id != &workspace_id)
            // A workspace ID that is still someone's local key means the
            // binding storing it resolves — not deadlocked, don't rewrite.
            .filter(|_| !accounts.contains_key(&workspace_id))
            .map(|local_id| (workspace_id, local_id))
    })
    .collect()
```

This loses zero deadlock coverage: a deadlocked binding stores an ID absent from the store, which by definition can't collide with a live key. Please also add a regression test using the true quarantined shape (insert a `CodexAccountData` with `chatgpt_account_id: None` directly into the accounts map, the way the v1 quarantine test builds it) so the shape mismatch can't recur.

Fix that one item and this is good to merge — happy to re-review quickly.

**Comment by allenxu09** [comment]
Addressed the remaining quarantined-v1 P1 in 2a9ea20e. Alias generation now refuses to rewrite any workspace ID that is still a live local account key, and the regression now inserts the real quarantined shape with chatgpt_account_id: None directly into the account map. Verified with cargo test --manifest-path src-tauri/Cargo.toml binding (8 passed), cargo fmt --manifest-path src-tauri/Cargo.toml --all -- --check, and git diff --check. I have also formally requested @farion1231 for re-review.

**Review by SaladDay — COMMENTED** [review]
I found one remaining cross-user binding risk in the current alias repair.

A legacy provider can still store workspace W after user A local record is gone. If user B later signs in to the same Team workspace and is the only ready account for W, legacy_binding_account_id_aliases treats B as the unique candidate and silently rewrites A provider to B local account. The accounts.contains_key(W) veto only protects the case where A legacy record is still present; it cannot prove who an absent binding originally belonged to.

Provider metadata has no historical stable user identity, so workspace uniqueness among current records is not enough to authorize an automatic rewrite. I recommend keeping unresolved bindings unresolved and requiring an explicit user choice, or only auto-repairing when a historical stable identity can be matched. The switch-away deadlock can be fixed independently without guessing a replacement account, as in #7060.

Because this can bind one Team user provider to another user credentials, I do not think the current PR is safe to merge as written.

**Comment by allenxu09** [comment]
Closing after the latest review identified that the legacy workspace ID alone cannot prove which Team user owned a deleted account. Any automatic alias rewrite can therefore bind a provider to another user credentials. The switch-away deadlock is being handled separately in #7060; a future binding repair should require explicit user selection or a persisted stable user identity. Keeping the branch for reference.
